### PR TITLE
Rename Shift to Raise according to Slack Poll, and add some initial docs

### PR DESCRIPTION
**Pull request by nomisRev**
After a poll on Slack, the new name `Raise` was most voted on.

There was some concern that `Raise` implied `Throwable`, but this seems to be an issue that is more widespread in the FP community. In Haskell & Scala the related type-classes typically also mention `Error`, `Except`, or similar and I think this is an issue we can easily overcome by properly explaining it in the docs. Therefore I added an initial draft of the documentation inside this PR. (cc\\ @serras perhaps you have some interesting insights on this topic?)

This PR is the place for final discussions, feedback and concerns before we settle on the name `Raise` for Arrow 2.0.

https://kotlinlang.slack.com/archives/C5UPMM0A0/p1661170633105679
Screenshot 2022-09-24 at 12 24 56

### CODE REVIEW

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+   * Note that this functions can currently not be _inline_ without Context Receivers,
+   * but you can rely on [recover] if you want to run any suspend code in your error handler.
+   *
+   * ```kotlin
+   * fun main() {
+   *   val failure: Result<Int> = Result.failure(RuntimeException("Boom!"))
+   *   either {
+   *     recover({ failure.bind() }) { failure: Throwable ->
+   *       delay(10)
+   *       shift("Something bad happened: ${failure.message}")
```
Comment (serras) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:128]: Missing change to `raise`

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+   * Note that this functions can currently not be _inline_ without Context Receivers,
+   * but you can rely on [recover] if you want to run any suspend code in your error handler.
+   *
+   * ```kotlin
+   * fun main() {
+   *   val empty: Option<Int> = None
+   *   either {
+   *     recover({ empty.bind() }) { failure: None ->
+   *       delay(10)
+   *       shift("Something bad happened: ${failure.message}")
```
[the remark this answers is not in the document]
Comment (serras) [reply arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:152]:
```suggestion
   *       raise("Something bad happened: ${failure.message}")
 ```

**Review by serras — APPROVED** [review]
Looks good. I like your approach of talking about "logic failures", to avoid focusing too much on validation, or exceptions, or anything different.

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmName
+
+@DslMarker
+public annotation class EffectDSL
+
+/**
+ * The [Raise] DSL allows you to work with _logical failures_ of type [R].
```
Comment (i-walker) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/Raise.kt:20]: 👏🏾

**Review by i-walker — APPROVED** [review]
Thanks to both @nomisRev and @serras !!

**Review by raulraja — APPROVED** [review]

**Comment by nomisRev** [comment]
Going to merge this PR, feedback is still welcome on this closed PR 🙏

**Pull request merged**